src/NServiceBus.Core/Hosting/Helpers/AssemblyScanner.cs

```diff
@@ -111,10 +111,12 @@ static List<Type> GetHandlerMessageTypes(List<Type> list)
                     {
                         var messageType = @interface.GetGenericArguments()[0];
                         foundMessageTypes.Add(messageType);
+
+                        foundMessageTypes.AddRange(messageType.Assembly.GetTypes().Where(t=> messageType.IsAssignableFrom(t)));
```

Comment:
How about we do the following (just an idea, haven't fully thought it through):

Currently the scanner ignores all assembly that don't reference NServiceBus directly. We could introduce an internal mode that allows us to bypass this and then do two passes.

Go through the assemblies and ignore everything that doesn't reference NSB.
Get the message types that are explicitly handled
Pass those assemblies into the scanner again and disable the "ignore not reference NSB mode"

But unfortunately this would also have severe impact on perf